src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,131 @@
+            // casting here to separate concerns and avoid 64 bit scalar deserialization.
+            interval as u64,
+            limit as usize,
+        )
+    }
+}
+
+/// Creates a log events stream based on component names, and a provided interval. Will emit
+/// control messages that bubble up the application if the sink goes away. The stream contains
+/// all matching events; filtering should be done at the caller level.
+fn create_log_events_stream(
+    control_tx: ControlSender,
+    component_names: &[String],
+    interval: u64,
+    limit: usize,
+) -> impl Stream<Item = Vec<LogEventResult>> {
+    let (tx, mut rx) = mpsc::channel(limit);
+    let (mut log_tx, log_rx) = mpsc::channel::<Vec<LogEventResult>>(10);
+
+    let tap_sink = TapSink::new(&component_names, tx);
+
+    tokio::spawn(async move {
+        // The tap controller is scoped to the stream. When it's dropped, it bubbles a control
+        // message up to the signal handler to remove the ad hoc sinks from topology.
+        let _control = TapController::new(control_tx, tap_sink);
+
+        let mut interval = time::interval(time::Duration::from_millis(interval));
+        let mut results: Vec<LogEventResult> = vec![];
+
+        loop {
+            select! {
+                // Process `TapResults`s. A tap result could contain a `LogEvent` or an error; if
+                // we get an error, the subscription is dropped.
+                Some(tap) = rx.next() => {
+                    results.push(tap.into());
+                }
+                _ = interval.tick() => {
+                    // If there are any existing results after the interval tick, emit.
+                    if !results.is_empty() {
+                        let results_len = results.len();
+
+                        // Events are 'sampled' up to the maximum 'limit', per an even
+                        // distribution of all events captured over the interval. We enumerate
+                        // chunks here to ensure the very last event is always returned when
+                        // limit > 1.
+                        let results = results
+                            .drain(..)
+                            .chunks(results_len.checked_div(limit).unwrap_or(1).max(1))
+                            .into_iter()
+                            .enumerate()
+                            .flat_map(|(i, chunk)| {
+                                let mut chunk = chunk.collect_vec();
+                                if matches!(i.checked_sub(1), Some(i) if i > 0 && i == results_len - 1) {
+                                    chunk = chunk.into_iter().rev().collect_vec();
+                                }
+                                chunk.into_iter().take(1)
+                            })
+                            .take(limit)
+                            .collect();
+
+                        let _ = log_tx.send(results).await;
```

Comment:
Do we care about handling the error here?

Reply:
I'm not sure there's a lot we could do about an error at that point. It would mean `log_rx` has gone away, which likely just means the subscription has been closed.

Reply:
Should we break out of the `loop` in that case?

Reply:
Yeah, you're right - this should be a hard break. I was thinking this was still wrapped in a `stream!` which would have gone away when the subscription ended, but that's no longer the case. I'll fix.